tests/integration/targets/terraform/tasks/test_provider_upgrade.yml

```diff
@@ -0,0 +1,23 @@
+---
+
+- name: Output terraform provider test project
+  ansible.builtin.template:
+    src: templates/provider_test/main.tf.j2
+    dest: "{{ terraform_project_dir }}/{{ tf_provider['name'] }}/main.tf"
+    force: yes
+  register: terraform_provider_hcl
+
+# The purpose of this task is to init terraform multiple times with different provider module
+# versions, so that we can verify that provider upgrades during init work as intended.
+
+- name: Init Terraform configuration with pinned provider version
+  community.general.terraform:
+    project_path: "{{ terraform_provider_hcl.dest | dirname }}"
+    binary_path: "{{ remote_tmp_dir }}/terraform"
```

Comment:
How will this work when terraform wasn't downloaded by this test?

Reply:
It will fail. If you have suggestions on how you guys normally handle outside dependencies, let me know.

Reply:
I'd probably use `set_fact` to create a new fact which points to the binary, something like
```
- set_fact:
    terraform_path: >-
      {{ remote_tmp_dir ~ '/terraform' if terraform_version_installed is not defined or terrform_version_installed != terraform_version else terraform_binary_path.stdout }}
```

Reply:
Just so I understand, the desired outcome here is that the end user can specify a path to an existing terraform binary, without the test having to download the binaries every run? I am trying to understand what we're solving for by setting the fact. Thanks!

Reply:
@felixfontein I think maybe I missed what you were asking initally. The playbook checks if the terraform binary is present in $PATH. If not, or if the version specified in `terraform_version` does not match what's present, the download will be initiated. But I see where the actual init task has a hardcoded binary_path of `{{ remote_tmp_dir }}/terraform`, which could be different if terraform was present in $PATH. So I will change it to coalesce the output of terraform detected in path, and then one installed, in that order.

Reply:
OK, I think we've got it :). Here's how it works now.

If `existing_terraform_path` is specified, it will be used by the playbook, and all version check/download tasks will be skipped.

If `existing_terraform_path` is empty or NOT defined, the check for terraform in path will still happen. Then the version check on that binary, if it exists.

If no terraform is found in path OR the terraform in path is not the right version (based on `terraform_version`), the download tasks will happen.

Finally, the `init terraform` task now references a variable for the binary path, in that order: user-defined path, path from the existing binary in $PATH, and finally, the one from temp dir that was downloaded.

Reply:
ansible-test does not allow generally to pass in user input, so supporting `existing_terraform_path` passed in doesn't really make sense. It suffices IMO to support having the binary in $PATH and downloading it.

Reply:
See my other comment below, this was reverted.